core/shared/src/main/scala/fs2/interop/flow/StreamSubscriber.scala

```diff
@@ -240,17 +239,21 @@ private[flow] final class StreamSubscriber[F[_], A] private (
 
         case WaitingOnUpstream(cb, s) =>
           Terminal -> run {
+            // We do the updates here,
+            // to ensure they happen after we have secured the state.
             if (canceled) {
               s.cancel()
+              inOnNextLoop = false
+              buffer = null
               cb.apply(Right(None))
-            } else if (index == 0) {
+            } else if (buffer eq null) {
```

Comment:
Rather than relying on the `index`, we simply check if the `buffer` hasn't been initialized.